src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,221 @@
+using System.Linq;
+using System.Windows.Controls;
+using Microsoft.CodeAnalysis.Editor.Implementation.Adornments;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.CodeAnalysis.Shared.TestHooks;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Adornments;
+using Microsoft.VisualStudio.Text.Classification;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+using Microsoft.VisualStudio.Text.Tagging;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Editor.InlineDiagnostics
+{
+    internal class InlineDiagnosticsAdornmentManager : AbstractAdornmentManager<InlineDiagnosticsTag>
+    {
+        private readonly IClassificationTypeRegistryService _classificationRegistryService;
+        private readonly IClassificationFormatMap _formatMap;
+
+        public InlineDiagnosticsAdornmentManager(
+            IThreadingContext threadingContext, IWpfTextView textView, IViewTagAggregatorFactoryService tagAggregatorFactoryService,
+            IAsynchronousOperationListener asyncListener, string adornmentLayerName,
+            IClassificationFormatMapService classificationFormatMapService,
+            IClassificationTypeRegistryService classificationTypeRegistryService)
+            : base(threadingContext, textView, tagAggregatorFactoryService, asyncListener, adornmentLayerName)
+        {
+            _classificationRegistryService = classificationTypeRegistryService;
+            _formatMap = classificationFormatMapService.GetClassificationFormatMap(textView);
+            _formatMap.ClassificationFormatMappingChanged += OnClassificationFormatMappingChanged;
+            TextView.ViewportWidthChanged += TextView_ViewportWidthChanged;
+        }
+
+        /// <summary>
+        /// Need to remove the tags if they intersect with the editor view, but only if the option
+        /// to place the tags at the end of the editor is selected.
+        /// </summary>
+        private void TextView_ViewportWidthChanged(object sender, EventArgs e)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            if (AdornmentLayer is null)
+            {
+                return;
+            }
+
+            var sourceContainer = TextView.TextBuffer.AsTextContainer();
+            if (sourceContainer is null)
+            {
+                AdornmentLayer.RemoveAllAdornments();
+                return;
+            }
+
+            if (!Workspace.TryGetWorkspace(sourceContainer, out var workspace))
+            {
+                AdornmentLayer.RemoveAllAdornments();
+                return;
+            }
+
+            var document = sourceContainer.GetOpenDocumentInCurrentContext();
+            if (document is null)
+            {
+                AdornmentLayer.RemoveAllAdornments();
+                return;
```

Comment:
so, a cleaner pattern for this is to have soemthing like: 

```c#
void TextView_ViewportWidthChanged()
{
  if (!TryTextView_ViewportWidthChanged())
    AdornmentLayer.RemoveAllAdornments();
}
```

that way you don't have to remember to annotate all failure return paths.

Reply:
(you can also put a nice comment here stating hwat's going on.